### DISCUSSION TITLE
Appetite for ".proto" file parser?

### DISCUSSION THREAD

**Discussion by jeffparsons** [tokio-rs/prost#888] [Ideas]
Related: https://github.com/tokio-rs/prost/discussions/486

I have two questions in mind:

- Would the Prost maintainers be interested in having a proto file parser in-tree, or is this too far outside the scope of the project?
- Would many Prost users (or just anyone working at the intersection of Rust and Protobufs, really) be interested in having a standalone proto file parser crate?

My primary use case involves custom code generation for several different languages (and docs, etc.). Currently this is achieved using a combination of protoc plugins (written in C++) and a fork of ruby-protoc, all glued together with some rather awkward Rake tasks.

I imagine using something like Chumsky and exposing a library interface and CLI tool. I would aim for bug-for-bug compatibility with the official protoc to the extent that is practicable.

Does this idea appeal to anyone?

**Comment by andrewhickman** [comment]
I've developed a crate for this: [protox](https://crates.io/crates/protox) which I think pretty much matches what you describe. It aims for equivalent output to protoc on all valid protobuf files.

I also pulled out the parser into its own crate [protox-parse](https://crates.io/crates/protox-parse/0.4.1) purely because the API is so small.
